File: src/main/java/com/kingdom/util/specialaction/SalvationSpecialActionHandler.java
package com.kingdom.util.specialaction;

import com.kingdom.model.*;
import com.kingdom.util.KingdomUtil;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class SalvationSpecialActionHandler {
    public static IncompleteCard handleSpecialAction(Game game, Card card) {

        IncompleteCard incompleteCard = null;
        Player player = game.getCurrentPlayer();

        if (card.getName().equals("Alms")) {
            CardAction cardAction = new CardAction(CardAction.TYPE_TRASH_CARDS_FROM_HAND);
            cardAction.setDeck(Deck.Salvation);
            cardAction.setCardName(card.getName());
            cardAction.setButtonValue("Done");
            cardAction.setNumCards(1);
            cardAction.setInstructions("Select a treasure card to trash.");
            cardAction.setCards(KingdomUtil.INSTANCE.uniqueCardList(player.getTreasureCards()));
            if (cardAction.getCards().size() > 0) {
                game.setPlayerCardAction(player, cardAction);
            } else {
                game.setPlayerInfoDialog(player, InfoDialog.Companion.getInfoDialog("You don't have any treasure cards to trash."));
            }
        } else if (card.getName().equals("Archbishop")) {
            CardAction cardAction = new CardAction(CardAction.TYPE_CHOICES);
            cardAction.setDeck(Deck.Salvation);
            cardAction.setCardName(card.getName());
            cardAction.setInstructions("Choose one: +2 Actions; remove 1 sin; or each other player gains 1 sin.");
            cardAction.getChoices().add(new CardActionChoice("+2 Actions", "actions"));
            cardAction.getChoices().add(new CardActionChoice("Remove 1 Sin", "remove"));
            cardAction.getChoices().add(new CardActionChoice("Others +1 Sin", "sins"));
            game.setPlayerCardAction(player, cardAction);
        } else if (card.getName().equals("Assassin")) {
            incompleteCard = new MultiPlayerIncompleteCard(card.getName(), game, false);
            for (Player otherPlayer : game.getPlayers()) {
                if (otherPlayer.getUserId() != game.getCurrentPlayerId()) {
                    if (game.isCheckEnchantedPalace() && game.revealedEnchantedPalace(otherPlayer.getUserId())) {
                        incompleteCard.setPlayerActionCompleted(otherPlayer.getUserId());
                        game.addHistory(otherPlayer.getUsername(), " revealed an ", KingdomUtil.INSTANCE.getWordWithBackgroundColor("Enchanted Palace", Card.VICTORY_AND_REACTION_IMAGE));
                    } else if (!otherPlayer.hasMoat() && !otherPlayer.hasLighthouse()) {
                        CardAction cardAction = new CardAction(CardAction.TYPE_TRASH_CARDS_FROM_HAND);
                        cardAction.setDeck(Deck.Salvation);
                        cardAction.setCardName(card.getName());
                        cardAction.setButtonValue("Done");
                        cardAction.setNumCards(1);
                        cardAction.setInstructions("Select an attack card to trash.");
                        for (Card c : otherPlayer.getActionCards()) {
                            if (c.isAttack()) {
                                cardAction.getCards().add(c);
                            }
                        }
                        if (!cardAction.getCards().isEmpty()) {
                            game.setPlayerCardAction(otherPlayer, cardAction);
                        } else {
                            incompleteCard.setPlayerActionCompleted(otherPlayer.getUserId());
                            game.addHistory(otherPlayer.getUsername(), " did not have any attack cards");
                        }
                    } else {
                        incompleteCard.setPlayerActionCompleted(otherPlayer.getUserId());
                        if (otherPlayer.hasLighthouse()) {
                            game.addHistory(otherPlayer.getUsername(), " had a ", KingdomUtil.INSTANCE.getWordWithBackgroundColor("Lighthouse", Card.ACTION_DURATION_COLOR));
                        } else {
                            game.addHistory(otherPlayer.getUsername(), " had a ", KingdomUtil.INSTANCE.getWordWithBackgroundColor("Moat", Card.ACTION_REACTION_COLOR));
                        }
                    }
                }
            }
            incompleteCard.allActionsSet();
        } else if (card.getName().equals("Baptistry")) {
            CardAction cardAction = new CardAction(CardAction.TYPE_CHOOSE_CARDS);
            cardAction.setDeck(Deck.Salvation);
            cardAction.setCardName(card.getName());
            cardAction.setButtonValue("Done");
            cardAction.setNumCards(1);
            cardAction.setInstructions("Choose a card.");
            cardAction.getCards().addAll(game.getCardMap().values());
            if (cardAction.getCards().size() > 0) {
                game.setPlayerCardAction(player, cardAction);
            }
        } else if (card.getName().equals("Catacombs")) {
            List<Card> cards = new ArrayList<Card>();
            Collections.shuffle(player.getDiscard());
            game.refreshDiscard(player);
            if (player.getDiscard().size() <= 4) {
                cards.addAll(player.getDiscard());
            } else {
                cards.addAll(player.getDiscard().subList(player.getDiscard().size() - 4, player.getDiscard().size()));
            }
            if (cards.isEmpty()) {
                game.setPlayerInfoDialog(player, InfoDialog.Companion.getInfoDialog("Your discard pile was empty."));
            } else {
                game.addHistory("Catacombs revealed ", KingdomUtil.INSTANCE.groupCards(cards, true));
                if (cards.size() == 1) {
                    game.playerGainedCardToHand(player, cards.get(0), false);
                    player.getDiscard().removeLastOccurrence(cards.get(0));
                } else if (cards.size() > 0) {
                    CardAction cardAction = new CardAction(CardAction.TYPE_CHOOSE_CARDS);
                    cardAction.setDeck(Deck.Salvation);
                    cardAction.setCardName(card.getName());
                    cardAction.setButtonValue("Done");
                    cardAction.setNumCards(1);
                    cardAction.setCards(KingdomUtil.INSTANCE.uniqueCardList(cards));
                    cardAction.setInstructions("Select a card to gain from your discard pile.");
                    game.setPlayerCardAction(player, cardAction);
                }
            }
        } else if (card.getName().equals("Edict")) {
            CardAction cardAction = new CardAction(CardAction.TYPE_CHOOSE_CARDS);
            cardAction.setDeck(Deck.Salvation);
            cardAction.setCardName(card.getName());
            cardAction.setButtonValue("Done");
            cardAction.setNumCards(1);
            cardAction.setInstructions("Choose an action card for the Edict.");
            for (Card c : game.getCardMap().values()) {
                if (c.isAction()) {
                    cardAction.getCards().add(c);
                }
            }
            game.setPlayerCardAction(player, cardAction);
        } else if (card.getName().equals("Graverobber")) {
            incompleteCard = new SinglePlayerIncompleteCard(card.getName(), game);
            for (Player otherPlayer : game.getPlayers()) {
                if (otherPlayer.getUserId() != game.getCurrentPlayerId()) {
                    if (game.isCheckEnchantedPalace() && game.revealedEnchantedPalace(otherPlayer.getUserId())) {
                        game.addHistory(otherPlayer.getUsername(), " revealed an ", KingdomUtil.INSTANCE.getWordWithBackgroundColor("Enchanted Palace", Card.VICTORY_AND_REACTION_IMAGE));
                    } else if (!otherPlayer.hasMoat() && !otherPlayer.hasLighthouse() && !otherPlayer.getDiscard().isEmpty()) {
                        int randomIndex1 = 0;
                        int randomIndex2 = -1;
                        if (otherPlayer.getDiscard().size() > 1) {
                            randomIndex1 = KingdomUtil.INSTANCE.getRandomNumber(0, otherPlayer.getDiscard().size() - 1);
                            randomIndex2 = KingdomUtil.INSTANCE.getRandomNumber(0, otherPlayer.getDiscard().size() - 1);
                            while (randomIndex2 == randomIndex1) {
                                randomIndex2 = KingdomUtil.INSTANCE.getRandomNumber(0, otherPlayer.getDiscard().size() - 1);
                            }
                        }
                        for (int i = 1; i <= 2; i++) {
                            Card revealedCard = null;
                            int index = 0;
                            if (i == 1) {
                                revealedCard = otherPlayer.getDiscard().get(randomIndex1);
                                index = randomIndex1;
                            } else if (i == 2 && randomIndex2 >= 0) {
                                revealedCard = otherPlayer.getDiscard().get(randomIndex2);
                                index = randomIndex2;
                            }
                            if (revealedCard != null) {
                                CardAction nextCardAction;
                                if (revealedCard.isTreasure()) {
                                    nextCardAction = new CardAction(CardAction.TYPE_YES_NO);
                                    nextCardAction.setInstructions("The graverobber revealed this treasure card from " + otherPlayer.getUsername() + "'s discard pile. Do you want to gain this card?");
                                } else {
                                    nextCardAction = new CardAction(CardAction.TYPE_CHOOSE_CARDS);
                                    nextCardAction.setNumCards(0);
                                    nextCardAction.setButtonValue("Continue");
                                    nextCardAction.setInstructions("The graverobber revealed this non-treasure card from " + otherPlayer.getUsername() + "'s discard pile. Click Continue.");
                                }
                                nextCardAction.setCardId(index);
                                nextCardAction.setDeck(Deck.Salvation);
                                nextCardAction.setPlayerId(otherPlayer.getUserId());
                                nextCardAction.setCardName(card.getName());
                                nextCardAction.getCards().add(revealedCard);
                                incompleteCard.getExtraCardActions().add(nextCardAction);
                            }
                        }
                    } else {
                        if (otherPlayer.hasLighthouse()) {
                            game.addHistory(otherPlayer.getUsername(), " had a ", KingdomUtil.INSTANCE.getWordWithBackgroundColor("Lighthouse", Card.ACTION_DURATION_COLOR));
                        } else if (otherPlayer.hasMoat()) {
                            game.addHistory(otherPlayer.getUsername(), " had a ", KingdomUtil.INSTANCE.getWordWithBackgroundColor("Moat", Card.ACTION_REACTION_COLOR));
                        } else {
                            game.addHistory(otherPlayer.getUsername(), "'s discard pile was empty");
                        }
                    }
                }
            }
            if (!incompleteCard.getExtraCardActions().isEmpty()) {
                CardAction cardAction = incompleteCard.getExtraCardActions().remove();
                cardAction.setDeck(Deck.Salvation);
                game.setPlayerCardAction(player, cardAction);
            }
        } else if (card.getName().equals("Indulgence")) {
            for (Player otherPlayer : game.getPlayers()) {
                if (otherPlayer.getUserId() != game.getCurrentPlayerId()) {
                    otherPlayer.addSins(-1);
                }
            }
            game.refreshAllPlayersPlayers();
            game.addHistory("All other players removed one sin");
        } else if (card.getName().equals("Inquisitor")) {
            for (Player otherPlayer : game.getPlayers()) {
                if (otherPlayer.getUserId() != game.getCurrentPlayerId()) {
                    if (game.isCheckEnchantedPalace() && game.revealedEnchantedPalace(otherPlayer.getUserId())) {
                        game.addHistory(otherPlayer.getUsername(), " revealed an ", KingdomUtil.INSTANCE.getWordWithBackgroundColor("Enchanted Palace", Card.VICTORY_AND_REACTION_IMAGE));
                    } else if (!otherPlayer.hasMoat() && !otherPlayer.hasLighthouse()) {
                        if (otherPlayer.getSins() >= 2) {
                            int cursesInSupply = game.getSupply().get(Card.CURSE_ID);
                            if (cursesInSupply >= 2) {
                                otherPlayer.addSins(-2);
                                game.addHistory(otherPlayer.getUsername(), " removed 2 sins and gained 2 ", KingdomUtil.INSTANCE.getWordWithBackgroundColor("Curse", Card.CURSE_COLOR), " cards from the Inquisitor");
                                game.playerGainedCard(otherPlayer, game.getCurseCard());
                                game.playerGainedCard(otherPlayer, game.getCurseCard());
                            } else if (cursesInSupply == 1) {
                                otherPlayer.addSins(-1);
                                game.addHistory(otherPlayer.getUsername(), " removed 1 sin and gained 1 ", KingdomUtil.INSTANCE.getWordWithBackgroundColor("Curse", Card.CURSE_COLOR), " card from the Inquisitor");
                                game.playerGainedCard(otherPlayer, game.getCurseCard());
                            } else {
                                otherPlayer.addSins(1);
                                game.addHistory(otherPlayer.getUsername(), " gained a sin from the Inquisitor");
                            }
                        } else {
                            otherPlayer.addSins(1);
                            game.addHistory(otherPlayer.getUsername(), " gained a sin from the Inquisitor");
                        }
                        game.refreshHandArea(otherPlayer);
                    } else {
                        if (otherPlayer.hasLighthouse()) {
                            game.addHistory(otherPlayer.getUsername(), " had a ", KingdomUtil.INSTANCE.getWordWithBackgroundColor("Lighthouse", Card.ACTION_DURATION_COLOR));
                        } else {
                            game.addHistory(otherPlayer.getUsername(), " had a ", KingdomUtil.INSTANCE.getWordWithBackgroundColor("Moat", Card.ACTION_REACTION_COLOR));
                        }
                    }
                }
            }
            game.refreshAllPlayersPlayers();
        } else if (card.getName().equals("Mendicant")) {
            CardAction cardAction = new CardAction(CardAction.TYPE_CHOOSE_UP_TO);
            cardAction.setDeck(Deck.Salvation);
            cardAction.setCardName(card.getName());
            cardAction.setButtonValue("Done");
            cardAction.setNumCards(1);
            cardAction.setInstructions("Select a card to gain from the trash pile and then click Done, or just click Done if you do not want to gain a card.");
            cardAction.setCards(KingdomUtil.INSTANCE.uniqueCardList(game.getTrashedCards()));
            if (cardAction.getCards().size() > 0) {
                game.setPlayerCardAction(player, cardAction);
            } else {
                game.setPlayerInfoDialog(player, InfoDialog.Companion.getInfoDialog("The trash pile is empty."));
            }
        } else if (card.getName().equals("Scriptorium")) {
            CardAction cardAction = new CardAction(CardAction.TYPE_DISCARD_FROM_HAND);
            cardAction.setDeck(Deck.Salvation);
            cardAction.setCardName(card.getName());
            cardAction.setButtonValue("Done");
            cardAction.setNumCards(1);
            cardAction.setInstructions("Chose an action card to discard from your hand.");
            cardAction.setCards(KingdomUtil.INSTANCE.uniqueCardList(player.getActionCards()));
            if (cardAction.getCards().size() > 0) {
                game.setPlayerCardAction(player, cardAction);
            } else {
                game.setPlayerInfoDialog(player, InfoDialog.Companion.getInfoDialog("You do not have any action cards in your hand."));
            }
        }

        return incompleteCard;
    }
}

File: src/main/kotlin/com/kingdom/util/specialaction/SalvationSpecialActionHandler.kt
package com.kingdom.util.specialaction

import com.kingdom.model.*
import com.kingdom.model.cards.Card
import com.kingdom.model.cards.CardColor
import com.kingdom.model.cards.Deck
import com.kingdom.model.cards.supply.Curse
import com.kingdom.util.KingdomUtil

import java.util.ArrayList
import java.util.Collections

object SalvationSpecialActionHandler {
    fun handleSpecialAction(game: OldGame, card: Card): IncompleteCard? {

        var incompleteCard: IncompleteCard? = null
        val player = game.currentPlayer

        when (card.name) {
            "Alms" -> {
                val cardAction = OldCardAction(OldCardAction.TYPE_TRASH_CARDS_FROM_HAND)
                cardAction.deck = Deck.Salvation
                cardAction.cardName = card.name
                cardAction.buttonValue = "Done"
                cardAction.numCards = 1
                cardAction.instructions = "Select a treasure card to trash."
                cardAction.cards = KingdomUtil.uniqueCardList(player!!.treasureCards)
                if (cardAction.cards.size > 0) {
                    game.setPlayerCardAction(player, cardAction)
                } else {
                    game.setPlayerInfoDialog(player, InfoDialog.getInfoDialog("You don't have any treasure cards to trash."))
                }
            }
            "Archbishop" -> {
                val cardAction = OldCardAction(OldCardAction.TYPE_CHOICES)
                cardAction.deck = Deck.Salvation
                cardAction.cardName = card.name
                cardAction.instructions = "Choose one: +2 Actions; remove 1 sin; or each other player gains 1 sin."
                cardAction.choices.add(CardActionChoice("+2 Actions", "actions"))
                cardAction.choices.add(CardActionChoice("Remove 1 Sin", "remove"))
                cardAction.choices.add(CardActionChoice("Others +1 Sin", "sins"))
                game.setPlayerCardAction(player!!, cardAction)
            }
            "Assassin" -> {
                incompleteCard = MultiPlayerIncompleteCard(card.name, game, false)
                for (otherPlayer in game.players) {
                    if (otherPlayer.userId != game.currentPlayerId) {
                        if (game.isCheckEnchantedPalace && game.revealedEnchantedPalace(otherPlayer.userId)) {
                            incompleteCard.setPlayerActionCompleted(otherPlayer.userId)
                            game.addHistory(otherPlayer.username, " revealed an ", KingdomUtil.getWordWithBackgroundColor("Enchanted Palace", CardColor.VictoryReaction))
                        } else if (!otherPlayer.hasMoat() && !otherPlayer.hasLighthouse()) {
                            val cardAction = OldCardAction(OldCardAction.TYPE_TRASH_CARDS_FROM_HAND)
                            cardAction.deck = Deck.Salvation
                            cardAction.cardName = card.name
                            cardAction.buttonValue = "Done"
                            cardAction.numCards = 1
                            cardAction.instructions = "Select an attack card to trash."
                            for (c in otherPlayer.actionCards) {
                                if (c.isAttack) {
                                    cardAction.cards.add(c)
                                }
                            }
                            if (!cardAction.cards.isEmpty()) {
                                game.setPlayerCardAction(otherPlayer, cardAction)
                            } else {
                                incompleteCard.setPlayerActionCompleted(otherPlayer.userId)
                                game.addHistory(otherPlayer.username, " did not have any attack cards")
                            }
                        } else {
                            incompleteCard.setPlayerActionCompleted(otherPlayer.userId)
                            if (otherPlayer.hasLighthouse()) {
                                game.addHistory(otherPlayer.username, " had a ", KingdomUtil.getWordWithBackgroundColor("Lighthouse", CardColor.ActionDuration))
                            } else {
                                game.addHistory(otherPlayer.username, " had a ", KingdomUtil.getWordWithBackgroundColor("Moat", CardColor.ActionReaction))
                            }
                        }
                    }
                }
                incompleteCard.allActionsSet()
            }
            "Baptistry" -> {
                val cardAction = OldCardAction(OldCardAction.TYPE_CHOOSE_CARDS)
                cardAction.deck = Deck.Salvation
                cardAction.cardName = card.name
                cardAction.buttonValue = "Done"
                cardAction.numCards = 1
                cardAction.instructions = "Choose a card."
                cardAction.cards.addAll(game.cardMap.values)
                if (cardAction.cards.size > 0) {
                    game.setPlayerCardAction(player!!, cardAction)
                }
            }
            "Catacombs" -> {
                val cards = ArrayList<Card>()
                Collections.shuffle(player!!.discard)
                game.refreshDiscard(player)
                if (player.discard.size <= 4) {
                    cards.addAll(player.discard)
                } else {
                    cards.addAll(player.discard.subList(player.discard.size - 4, player.discard.size))
                }
                if (cards.isEmpty()) {
                    game.setPlayerInfoDialog(player, InfoDialog.getInfoDialog("Your discard pile was empty."))
                } else {
                    game.addHistory("Catacombs revealed ", KingdomUtil.groupCards(cards, true))
                    if (cards.size == 1) {
                        game.playerGainedCardToHand(player, cards[0], false)
                        player.discard.removeLastOccurrence(cards[0])
                    } else if (cards.size > 0) {
                        val cardAction = OldCardAction(OldCardAction.TYPE_CHOOSE_CARDS)
                        cardAction.deck = Deck.Salvation
                        cardAction.cardName = card.name
                        cardAction.buttonValue = "Done"
                        cardAction.numCards = 1
                        cardAction.cards = KingdomUtil.uniqueCardList(cards)
                        cardAction.instructions = "Select a card to gain from your discard pile."
                        game.setPlayerCardAction(player, cardAction)
                    }
                }
            }
            "Edict" -> {
                val cardAction = OldCardAction(OldCardAction.TYPE_CHOOSE_CARDS)
                cardAction.deck = Deck.Salvation
                cardAction.cardName = card.name
                cardAction.buttonValue = "Done"
                cardAction.numCards = 1
                cardAction.instructions = "Choose an action card for the Edict."
                for (c in game.cardMap.values) {
                    if (c.isAction) {
                        cardAction.cards.add(c)
                    }
                }
                game.setPlayerCardAction(player!!, cardAction)
            }
            "Graverobber" -> {
                incompleteCard = SinglePlayerIncompleteCard(card.name, game)
                for (otherPlayer in game.players) {
                    if (otherPlayer.userId != game.currentPlayerId) {
                        if (game.isCheckEnchantedPalace && game.revealedEnchantedPalace(otherPlayer.userId)) {
                            game.addHistory(otherPlayer.username, " revealed an ", KingdomUtil.getWordWithBackgroundColor("Enchanted Palace", CardColor.VictoryReaction))
                        } else if (!otherPlayer.hasMoat() && !otherPlayer.hasLighthouse() && !otherPlayer.discard.isEmpty()) {
                            var randomIndex1 = 0
                            var randomIndex2 = -1
                            if (otherPlayer.discard.size > 1) {
                                randomIndex1 = KingdomUtil.getRandomNumber(0, otherPlayer.discard.size - 1)
                                randomIndex2 = KingdomUtil.getRandomNumber(0, otherPlayer.discard.size - 1)
                                while (randomIndex2 == randomIndex1) {
                                    randomIndex2 = KingdomUtil.getRandomNumber(0, otherPlayer.discard.size - 1)
                                }
                            }
                            for (i in 1..2) {
                                var revealedCard: Card? = null
                                var index = 0
                                if (i == 1) {
                                    revealedCard = otherPlayer.discard[randomIndex1]
                                    index = randomIndex1
                                } else if (i == 2 && randomIndex2 >= 0) {
                                    revealedCard = otherPlayer.discard[randomIndex2]
                                    index = randomIndex2
                                }
                                if (revealedCard != null) {
                                    val nextOldCardAction: OldCardAction
                                    if (revealedCard.isTreasure) {
                                        nextOldCardAction = OldCardAction(OldCardAction.TYPE_YES_NO)
                                        nextOldCardAction.instructions = "The graverobber revealed this treasure card from " + otherPlayer.username + "'s discard pile. Do you want to gain this card?"
                                    } else {
                                        nextOldCardAction = OldCardAction(OldCardAction.TYPE_CHOOSE_CARDS)
                                        nextOldCardAction.numCards = 0
                                        nextOldCardAction.buttonValue = "Continue"
                                        nextOldCardAction.instructions = "The graverobber revealed this non-treasure card from " + otherPlayer.username + "'s discard pile. Click Continue."
                                    }
                                    nextOldCardAction.deck = Deck.Salvation
                                    nextOldCardAction.playerId = otherPlayer.userId
                                    nextOldCardAction.cardName = card.name
                                    nextOldCardAction.cards.add(revealedCard)
                                    incompleteCard.extraOldCardActions.add(nextOldCardAction)
                                }
                            }
                        } else {
                            when {
                                otherPlayer.hasLighthouse() -> game.addHistory(otherPlayer.username, " had a ", KingdomUtil.getWordWithBackgroundColor("Lighthouse", CardColor.ActionDuration))
                                otherPlayer.hasMoat() -> game.addHistory(otherPlayer.username, " had a ", KingdomUtil.getWordWithBackgroundColor("Moat", CardColor.ActionReaction))
                                else -> game.addHistory(otherPlayer.username, "'s discard pile was empty")
                            }
                        }
                    }
                }
                if (!incompleteCard.extraOldCardActions.isEmpty()) {
                    val cardAction = incompleteCard.extraOldCardActions.remove()
                    cardAction.deck = Deck.Salvation
                    game.setPlayerCardAction(player!!, cardAction)
                }
            }
            "Indulgence" -> {
                for (otherPlayer in game.players) {
                    if (otherPlayer.userId != game.currentPlayerId) {
                        otherPlayer.addSins(-1)
                    }
                }
                game.refreshAllPlayersPlayers()
                game.addHistory("All other players removed one sin")
            }
            "Inquisitor" -> {
                for (otherPlayer in game.players) {
                    if (otherPlayer.userId != game.currentPlayerId) {
                        if (game.isCheckEnchantedPalace && game.revealedEnchantedPalace(otherPlayer.userId)) {
                            game.addHistory(otherPlayer.username, " revealed an ", KingdomUtil.getWordWithBackgroundColor("Enchanted Palace", CardColor.VictoryReaction))
                        } else if (!otherPlayer.hasMoat() && !otherPlayer.hasLighthouse()) {
                            if (otherPlayer.sins >= 2) {
                                val cursesInSupply = game.supply[Curse.NAME]!!
                                when {
                                    cursesInSupply >= 2 -> {
                                        otherPlayer.addSins(-2)
                                        game.addHistory(otherPlayer.username, " removed 2 sins and gained 2 ", KingdomUtil.getWordWithBackgroundColor("Curse", CardColor.Curse), " cards from the Inquisitor")
                                        game.playerGainedCard(otherPlayer, game.curseCard)
                                        game.playerGainedCard(otherPlayer, game.curseCard)
                                    }
                                    cursesInSupply == 1 -> {
                                        otherPlayer.addSins(-1)
                                        game.addHistory(otherPlayer.username, " removed 1 sin and gained 1 ", KingdomUtil.getWordWithBackgroundColor("Curse", CardColor.Curse), " card from the Inquisitor")
                                        game.playerGainedCard(otherPlayer, game.curseCard)
                                    }
                                    else -> {
                                        otherPlayer.addSins(1)
                                        game.addHistory(otherPlayer.username, " gained a sin from the Inquisitor")
                                    }
                                }
                            } else {
                                otherPlayer.addSins(1)
                                game.addHistory(otherPlayer.username, " gained a sin from the Inquisitor")
                            }
                            game.refreshHandArea(otherPlayer)
                        } else {
                            if (otherPlayer.hasLighthouse()) {
                                game.addHistory(otherPlayer.username, " had a ", KingdomUtil.getWordWithBackgroundColor("Lighthouse", CardColor.ActionDuration))
                            } else {
                                game.addHistory(otherPlayer.username, " had a ", KingdomUtil.getWordWithBackgroundColor("Moat", CardColor.ActionReaction))
                            }
                        }
                    }
                }
                game.refreshAllPlayersPlayers()
            }
            "Mendicant" -> {
                val cardAction = OldCardAction(OldCardAction.TYPE_CHOOSE_UP_TO)
                cardAction.deck = Deck.Salvation
                cardAction.cardName = card.name
                cardAction.buttonValue = "Done"
                cardAction.numCards = 1
                cardAction.instructions = "Select a card to gain from the trash pile and then click Done, or just click Done if you do not want to gain a card."
                cardAction.cards = KingdomUtil.uniqueCardList(game.trashedCards)
                if (cardAction.cards.size > 0) {
                    game.setPlayerCardAction(player!!, cardAction)
                } else {
                    game.setPlayerInfoDialog(player!!, InfoDialog.getInfoDialog("The trash pile is empty."))
                }
            }
            "Scriptorium" -> {
                val cardAction = OldCardAction(OldCardAction.TYPE_DISCARD_FROM_HAND)
                cardAction.deck = Deck.Salvation
                cardAction.cardName = card.name
                cardAction.buttonValue = "Done"
                cardAction.numCards = 1
                cardAction.instructions = "Chose an action card to discard from your hand."
                cardAction.cards = KingdomUtil.uniqueCardList(player!!.actionCards)
                if (cardAction.cards.size > 0) {
                    game.setPlayerCardAction(player, cardAction)
                } else {
                    game.setPlayerInfoDialog(player, InfoDialog.getInfoDialog("You do not have any action cards in your hand."))
                }
            }
        }

        return incompleteCard
    }
}
